src/super_gradients/training/utils/distributed_training_utils.py

```diff
@@ -171,30 +192,128 @@ def wait_for_the_master(local_rank: int):
+    """
+    If required, launch ddp subprocesses.
+    :param multi_gpu:    DDP, DP or Off
+    :param num_gpus:    Number of GPU's to use.
+    """
+    if not torch.cuda.is_available():
+        raise RuntimeError("Cuda device_config not found...")
+
+    if num_gpus in (-1, None):
+        if multi_gpu in (MultiGPUMode.OFF, MultiGPUMode.DATA_PARALLEL):
+            num_gpus = 1
+        elif multi_gpu in (MultiGPUMode.AUTO, MultiGPUMode.DISTRIBUTED_DATA_PARALLEL):
+            num_gpus = torch.cuda.device_count()
+
+    if multi_gpu == MultiGPUMode.AUTO:
+        if num_gpus > 1:
+            multi_gpu = MultiGPUMode.DISTRIBUTED_DATA_PARALLEL
+        else:
+            multi_gpu = MultiGPUMode.OFF
+
+    # Check that num_gpus is valid for specified multi_gpu
+    if multi_gpu in (MultiGPUMode.OFF, MultiGPUMode.DATA_PARALLEL):
+        if num_gpus != 1:
+            # TODO: Should we change to a warning, and set num_gpus=1 afterward ?
+            raise ValueError(f"You specified num_gpus={num_gpus} but it has not be 1 on when working with multi_gpu={multi_gpu}")
+    else:
         if num_gpus > torch.cuda.device_count():
+            # TODO: Should we change to a warning, and set num_gpus=torch.cuda.device_count() afterward ?
             raise ValueError(f"You specified num_gpus={num_gpus} but only {torch.cuda.device_count()} GPU's are available")
-        restart_script_with_ddp(num_gpus)
+
+    device_config.device = "cuda"
+    device_config.multi_gpu = multi_gpu
+
+    if multi_gpu == MultiGPUMode.DISTRIBUTED_DATA_PARALLEL:
+        if is_distributed():
+            initialize_ddp(local_rank=device_config.requested_rank)
+        else:
+            restart_script_with_ddp(num_gpus=num_gpus)
 
 
-def require_gpu_setup(gpu_mode: MultiGPUMode) -> bool:
-    """Check if the environment requires a setup in order to work with DDP."""
-    return (gpu_mode == MultiGPUMode.DISTRIBUTED_DATA_PARALLEL) and (not is_distributed())
+def initialize_ddp(local_rank: int):
+    """
+    Initialize Distributed Data Parallel
+
+    Important note: (1) in distributed training it is customary to specify learning rates and batch sizes per GPU.
+    Whatever learning rate and schedule you specify will be applied to the each GPU individually.
+    Since gradients are passed and summed (reduced) from all to all GPUs, the effective batch size is the
+    batch you specify times the number of GPUs. In the literature there are several "best practices" to set
+    learning rates and schedules for large batch sizes.
+    """
+
+    if local_rank > 0:
+        mute_current_process()
+
+    logger.info("Distributed training starting...")
+    if not torch.distributed.is_initialized():
+        backend = "gloo" if os.name == "nt" else "nccl"
+        torch.distributed.init_process_group(backend=backend, init_method="env://")
+    torch.cuda.set_device(local_rank)
+
+    assert device_config.requested_rank == get_local_rank()
```

Comment:
It is recommended to use assert only inside unit tests and use exceptions for runtime checks elsewhere.  assert can be tuned off by interpreter settings while runtime checks can't.